HPCC-17056 _stripWhitespaceFromStoredDataset not working in roxie

afishbeck opened the pull request:
Feature can be activated in 3 ways:
  1. #OPTION('stripWhitespaceFromStoredDataset', false); //default is true
  2. Query attribute <MyQuery stripWhitespaceFromStoredDataset='0'/>
  3. Query parameter:

<MyQuery>
  <value>101</value>
  <_stripWhitespaceFromStoredDataset>0</_stripWhitespaceFromStoredDataset>
</MyQuery>

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately

## Testing:
<!-- Please describe how this change has been tested.-->
Ran testsocket tests, soap tests, roxiepipe tests, with and without the stripWhitespace input parameter and/or hash option.
Added a new regression test to test this option.
You may want to merge the roxiepipe test I did previously to this branch?
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

afishbeck commented:
@richardkchapman I still have a bit more testing to do, but take a look.
You'll had to move some code around, mostly to be able to get the hash option in order, and maybe some other ordering issues.  Not sure if it's too extensive for a point release.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 119974d950eae67de902c69467f76d65aacce072
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.216 seconds |


Install hpccsystems-platform-community_6.2.7-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |


Regression test result:

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxie_keepwhitespace | Pass | Pass | Pass |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxiewhitespace | Excl. | Excl. | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 668837cf07fdcdcf2075b84bab280d1b020e98b7
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 90.064 seconds |


Install hpccsystems-platform-community_6.2.7-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |


Regression test result:

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxie_keepwhitespace | Pass | Pass | Pass |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxiewhitespace | Excl. | Excl. | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented on roxie/ccd/ccdprotocol.cpp:
```diff
@@ -1601,10 +1598,24 @@ class RoxieSocketWorker : public ProtocolQueryWorker
+        if (queryPT->hasProp("_stripWhitespaceFromStoredDataset"))
+        {
+            bool stripTag = queryPT->getPropBool("_stripWhitespaceFromStoredDataset");
+            bool stripFlag = options & ptr_ignoreWhiteSpace;
```
Does this need to be / would it be clearer as

  bool stripFlag = (options & ptr_ignoreWhiteSpace) != 0;

?

ghalliday replied:
Yes, visual studio will give a warning about implicit conversion to bool.

richardkchapman commented on roxie/ccd/ccdprotocol.cpp:
```diff
@@ -1601,10 +1598,24 @@ class RoxieSocketWorker : public ProtocolQueryWorker
+            bool stripFlag = options & ptr_ignoreWhiteSpace;
+            if (stripTag != stripFlag)
+            {
+                if (stripTag)
```
A comment here about why you had to parse twice and why there is no way to optimize that away might be handy to avoid future refactorings reintroducing the problem

richardkchapman commented:
The test case is good, but would be good to also have a test case that uses the #option, and (might be harder but maybe the roxiepipe test cases can give some clues) the attribute variants.

richardkchapman commented:
@ghalliday Please review

ghalliday commented on roxie/ccd/ccdprotocol.cpp:
```diff
@@ -1517,68 +1539,45 @@ class RoxieSocketWorker : public ProtocolQueryWorker
-            queryName.set(queryPT->queryName());
-            isRequest = false;
-            isRequestArray = false;
+            StringAttr tagName = queryPT->queryName();
```
I think you can use

    const char * tagName

ghalliday commented on roxie/ccd/ccdprotocol.cpp:
```diff
@@ -1350,17 +1354,29 @@ class QueryNameExtractor : implements IPTreeNotifyEvent, public CInterface
-    void extractName(const char *msg, const IContextLogger &logctx, const char *peer, unsigned port)
+    void extractName(HttpHelper &httpHelper, const char *msg, const IContextLogger &logctx, const char *peer, unsigned port)
     {
+        StringAttr urlName;
```
Can this also be a const char *?

ghalliday commented on roxie/ccd/ccdprotocol.cpp:
```diff
@@ -1714,10 +1727,9 @@ class RoxieSocketWorker : public ProtocolQueryWorker
 
-                IPropertyTree *root = queryPT;
-                skipProtocolRoot(queryPT, httpHelper, queryName, isRequest, isRequestArray);
+                //IPropertyTree *root = queryPT;
```
remove unused comment

afishbeck commented:
@richardkchapman In regards to the #option the half of the test that is published at setup (see 'roxie_keepwhitespace.ecl above') currently uses the #option as 'false' and the runtime query either does or doesn't over ride that.  I can add another published query that sets it to true (the default) if you think it necessary.  I added a setup query to 6.4 that I can just add it to if we migrate it to 6.2.8 (roxie_echo.ecl).

afishbeck commented:
@richardkchapman Updated, but I left one of Gavin's suggestions for a new JIRA, wanted to keep this one safe.

afishbeck committed:
HPCC-17056 _stripWhitespaceFromStoredDataset not working in roxie

Feature being fixed can be activated in 3 ways:
  1. #OPTION('stripWhitespaceFromStoredDataset', false); //default is true
  2. Query attribute <MyQuery stripWhitespaceFromStoredDataset='0'/>
  3. Query parameter:

<MyQuery>
  <value>101</value>
  <_stripWhitespaceFromStoredDataset>0</_stripWhitespaceFromStoredDataset>
</MyQuery>

Signed-off-by: Anthony Fishbeck <anthony.fishbeck@lexisnexis.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 2d0b4d74e588a5a934d64e735bc62838a24a99d5
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.8 seconds |


Install hpccsystems-platform-community_6.2.7-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |


Regression test result:

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxie_keepwhitespace | Pass | Pass | Pass |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxiewhitespace | Excl. | Excl. | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK

afishbeck commented:
@richardkchapman Squashed into single commit.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 9c833c2b730d2f985862bcf8151ba9611713a80c
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.111 seconds |


Install hpccsystems-platform-community_6.2.7-closedown0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |88 |88 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |


Regression test result:

| setup | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxie_keepwhitespace | Pass | Pass | Pass |
| setup | Pass | Pass | Pass |
| setup_fetch | Pass | Pass | Pass |
| setupdict | Pass | Pass | Pass |
| setupsearchindex | Pass | Pass | Pass |
| setupsq | Pass | Pass | Pass |
| setupwordindex | Pass | Pass | Pass |
| setupxml | Pass | Pass | Pass |
 

| test | hthor | thor | roxie |
| ----- | ----- | ----- | ---- |
| roxiewhitespace | Excl. | Excl. | Pass |
 
HPCC Stop: OK
HPCC Uninstall: OK
